cleanup dependencies & addons, and add an option to wipe intermediate build dirs

dhewg opened the pull request:
The new `get_install_dir()` and `sysroot` build flag from #4042 are used to get rid of most `get_build_dir()` calls. The idea is to seperate build from install dirs in a clean way, which eventually allows to add an optional autoremoval of build dirs to keep the size of the build folder down (set AUTOREMOVE=yes to enable):

Without autoremove | With autoremove
------ | ----
<img width="320" alt="before" src="https://user-images.githubusercontent.com/605548/71667621-2f32e480-2d66-11ea-8ba8-df641b7daa7f.png"> | <img width="320" alt="after" src="https://user-images.githubusercontent.com/605548/71667625-322dd500-2d66-11ea-80b2-c6c730eb0f5e.png">

Another advantage is that the build system already does post processing on install dirs, like python abi fixups and stripping binaries, which would have prevented the `i2c-tools` issue from #4070.

The general approach is:
* `get_build_dir()` cannot be used to access files created during the build process, only files which are part of the original tarball
* Use `PKG_BUILD_FLAGS="-sysroot"` instead of stubbing `make install`. `get_install_dir()` can then be used instead of fishing around in build dirs
* Always use `make install` to populate `install_pkg`. If packages don't have an install target, populate manually but keep the usual `/usr/bin`, `/usr/lib` etc. structure
* Use `get_install_dir()` where possible and `get_build_dir()` only where necessary. If `get_build_dir()` is required, so is `PKG_DEPENDS_UNPACK`
* If access to files is required, which should not be part of the image file system, copy/move them to a hidden directory under `install_pkg`

With that it's now possible to include addon dependency packages in the image, like `ADDITIONAL_PACKAGES="screen"`.

Noteworthy:
* Includes a patch for `pkg-config` to make it play nicely with `-sysroot` packages, and add `PKG_DEPENDS_CONFIG` on top of that to ease building against these
* Merge `rsync` and `rsyncx`
* Merge all `libX*` addon duplicates into the original

Notes:
* Allwinner/H6 image with a couple of addons runtime tested only
* build tested a few other targets where changes were made, but not every one
* most addons play along with autoremove, but not all. `scripts/create_addon all -docker -dotnet-runtime -game.libretro.uae -hyperion -librespot -syncthing -vdr-addon` works, some target specific addons like `adafruit-libraries` and `rpi-tools` need work too
* `-sysroot` is only used when `make install` was stubbed. There're alot addon dependencies which still throw their stuff in sysroot, which could be detected and build against by base packages if building an image after addons (which would yield a broken image)
* `PKG_NEED_UNPACK` is a misnomer and easy to confuse with `PKG_DEPENDS_UNPACK`. Maybe rename to e.g. `PKG_STAMP_FILES`?
* `PKG_DEPENDS_CONFIG` could use a better name. Right now it's just used for adding entries to `PKG_CONFIG_PATH`, but if this could be useful for similar stuff (cmake modules?), give it a generic name like e.g. `PKG_BUILDS_AGAINST`?

Kwiboo commented on packages/mediacenter/JsonSchemaBuilder/package.mk:
```diff
@@ -8,10 +8,9 @@ PKG_LICENSE="GPL"
 
-PKG_CMAKE_SCRIPT="$(get_build_dir $MEDIACENTER)/tools/depends/native/JsonSchemaBuilder/CMakeLists.txt"
+PKG_CMAKE_SCRIPT="$(get_build_dir kodi)/tools/depends/native/JsonSchemaBuilder/CMakeLists.txt"
 
```
Please use `$MEDIACENTER` instead of `kodi`, and `PKG_DEPENDS_UNPACK="kodi"` should be changed to `PKG_DEPENDS_UNPACK="$MEDIACENTER"`

dhewg replied:
done

Kwiboo commented on packages/mediacenter/TexturePacker/package.mk:
```diff
@@ -8,15 +8,14 @@ PKG_LICENSE="GPL"
 
 pre_configure_host() {
-  export CXXFLAGS="$CXXFLAGS -std=c++11 -DTARGET_POSIX -DTARGET_LINUX -D_LINUX -I$(get_build_dir $MEDIACENTER)/xbmc/platform/linux"
+  export CXXFLAGS="$CXXFLAGS -std=c++11 -DTARGET_POSIX -DTARGET_LINUX -D_LINUX -I$(get_build_dir kodi)/xbmc/platform/linux"
```
Same as above, please use `$MEDIACENTER` instead of `kodi`, and `PKG_DEPENDS_UNPACK="kodi"` should be changed to `PKG_DEPENDS_UNPACK="$MEDIACENTER"`

dhewg replied:
done

Kwiboo commented on projects/Rockchip/bootloader/release:
```diff
@@ -1,23 +1,23 @@
   DTB_PREFIX="${DEVICE/RK/rk}-"
-  LINUX_DTS_DIR=$(get_build_dir linux)/arch/$TARGET_KERNEL_ARCH/boot/dts
-  for dtb in $LINUX_DTS_DIR/$DTB_PREFIX*.dtb $LINUX_DTS_DIR/rockchip/$DTB_PREFIX*.dtb; do
+  for dtb in $SRCDIR/$DTB_PREFIX*.dtb $SRCDIR/rockchip/$DTB_PREFIX*.dtb; do
```
We should be able to replace this entire for block with:
`cp -a $SRCDIR/${DEVICE/RK/rk}-*.dtb $RELEASE_DIR/3rdparty/bootloader`

dhewg replied:
done

dhewg commented:
rebased due to conflicts

MilhouseVH commented:
Couple of quickies before I review the whole thing later, or over the weekend:

> * PKG_NEED_UNPACK is a misnomer and easy to confuse with PKG_DEPENDS_UNPACK. Maybe rename to e.g. PKG_STAMP_FILES?

I agree, it's not a great name but probably best to leave it as it is in this PR and we can bikeshed a new name on Slack or in a PR dedicated to the name change - that avoids taking the focus away from the rest of the changes in this PR.

> * PKG_DEPENDS_CONFIG could use a better name. Right now it's just used for adding entries to PKG_CONFIG_PATH, but if this could be useful for similar stuff (cmake modules?), give it a generic name like e.g. PKG_BUILDS_AGAINST?

To save time and avoid unnecessary chatter, I'd say keep as it is for now and we can bikeshed it with PKG_NEED_UNPACK in a future PR dedicated to finding a more suitable name. Changing the names should be pretty simple once we decide on replacement(s).

MilhouseVH commented on packages/linux-driver-addons/dvb/crazycat/package.mk:
```diff
@@ -7,8 +7,7 @@ PKG_SHA256="6b44a96d82c4a3e052864a995baceaede46b37c048c5718a6f62a009492d08ff"
 PKG_URL="https://bitbucket.org/CrazyCat/media_build/get/$PKG_VERSION.tar.gz"
-PKG_DEPENDS_TARGET="media_tree_cc"
-PKG_NEED_UNPACK="$(get_pkg_directory media_tree_cc)"
+PKG_DEPENDS_UNPACK="media_tree_cc"
```
I think want to keep `PKG_NEED_UNPACK`, the purpose of which is to automatically clean the `crazycat` package if the `media_tree_cc` package has changed.

With the changes above, we would unpack the new `media_tree_cc` package, but we wouldn't rebuild `crazycat` to use the new `media_tree_cc` changes.

MilhouseVH commented:
We should definitely document `$INSTALL/.noinstall` as that's new functionality and it isn't immediately obvious when it should be used/is needed etc.

MilhouseVH commented:
Everywhere `PKG_NEED_UNPACK` has been removed, we need to add it back, I think.

Alternatively, and which might be better, we iterate over `PKG_DEPENDS_UNPACK` and append the resulting `get_build_dir()`'s to `PKG_NEED_UNPACK` when sourcing the package.

MilhouseVH commented:
Never mind - just realised we already do exactly that (iterate and append). So yes, it's safe to get rid of `PKG_NEED_UNPACK` when using `PKG_DEPENDS_UNPACK` :smile:

MilhouseVH commented on packages/addons/tools/btrfs-progs/package.mk:
```diff
@@ -27,11 +28,7 @@ pre_configure_target() {
 addon() {
   mkdir -p $ADDON_BUILD/$PKG_ADDON_ID/bin/
-    cp -P $(get_build_dir btrfs-progs)/{btrfs,btrfsck,btrfstune,btrfs-zero-log,fsck.btrfs,mkfs.btrfs} $ADDON_BUILD/$PKG_ADDON_ID/bin
+    cp -P $PKG_BUILD/{btrfs,btrfsck,btrfstune,btrfs-zero-log,fsck.btrfs,mkfs.btrfs} $ADDON_BUILD/$PKG_ADDON_ID/bin
```
lol

MilhouseVH commented on packages/addons/addon-depends/chrome-depends/gtk3/package.mk:
```diff
@@ -9,7 +9,9 @@ PKG_LICENSE="LGPL"
 PKG_SITE="http://www.gtk.org/"
 PKG_URL="https://ftp.gnome.org/pub/gnome/sources/gtk+/${PKG_VERSION:0:4}/gtk+-$PKG_VERSION.tar.xz"
 PKG_DEPENDS_TARGET="toolchain at-spi2-atk atk cairo gdk-pixbuf glib libX11 libXi libXrandr libepoxy pango"
+PKG_DEPENDS_CONFIG="libXft pango gdk-pixbuf shared-mime-info"
```
Add `libXft` and `shared-mime-info` to `PKG_DEPENDS_TARGET`?

dhewg replied:
Same as mediainfo, those are 2nd level dependencies. Some mechanism to do all that automatically would be nice, but not exactly trivial. Do we really wanna add all these deps?

MilhouseVH replied:
No that's fine, in that case we just have to hope that the packages that currently have `libXft` and `shared-mime-info` as dependencies don't drop those dependencies in future, otherwise this package will break. :smile:

MilhouseVH commented on packages/addons/service/locale/package.mk:
```diff
@@ -16,13 +16,11 @@ PKG_ADDON_TYPE="xbmc.service"
 
-  mkdir -p "$ADDON_BUILD/$PKG_ADDON_ID/i18n"
-  cp -PR "$(get_build_dir glibc)/localedata/charmaps" \
-         "$(get_build_dir glibc)/localedata/locales" \
```
Don't we need charmaps and locales?

dhewg replied:
Those are supposed to come from the `usr/share/i18n` folder copied below, but that doesn't happen... will fix

MilhouseVH commented on packages/addons/service/proftpd/package.mk:
```diff
@@ -33,30 +34,26 @@ PKG_CONFIGURE_OPTS_TARGET="--enable-static \
-}
+                           --enable-largefile \
+                           install_user=$(id -u) \
+                           install_group=$(id -g)"
```
Is this going to be a problem with "reproducible builds"? What is this used for - does it affect the final binary?

dhewg replied:
This was required because we now use `make install` and upstream installs file for a specific user (chmod). But that requires root privs, except when the uid/gid matches the current user - which this just does to make `make install` succeed

MilhouseVH replied:
Ok thanks.

MilhouseVH commented on packages/addons/service/pcscd/package.mk:
```diff
@@ -24,12 +24,7 @@ addon() {
-  mkdir -p $ADDON_BUILD/$PKG_ADDON_ID/drivers/ifd-ccid.bundle/Contents/Linux/
-    cp -Pa $(get_build_dir ccid)/.$TARGET_NAME/src/.libs/libccid.so $ADDON_BUILD/$PKG_ADDON_ID/drivers/ifd-ccid.bundle/Contents/Linux/
-    cp -Pa $(get_build_dir ccid)/.$TARGET_NAME/src/Info.plist $ADDON_BUILD/$PKG_ADDON_ID/drivers/ifd-ccid.bundle/Contents
+  cp -a $(get_install_dir ccid)/storage/.kodi/addons/$PKG_ADDON_ID/drivers $ADDON_BUILD/$PKG_ADDON_ID
```
Is this source path correct?

dhewg replied:
I think so, yes:
```
install_pkg/ccid-1.4.28
install_pkg/ccid-1.4.28/.libreelec-package
install_pkg/ccid-1.4.28/etc
install_pkg/ccid-1.4.28/etc/reader.conf.d
install_pkg/ccid-1.4.28/etc/reader.conf.d/libccidtwin
install_pkg/ccid-1.4.28/storage
install_pkg/ccid-1.4.28/storage/.kodi
install_pkg/ccid-1.4.28/storage/.kodi/addons
install_pkg/ccid-1.4.28/storage/.kodi/addons/service.pcscd
install_pkg/ccid-1.4.28/storage/.kodi/addons/service.pcscd/drivers
install_pkg/ccid-1.4.28/storage/.kodi/addons/service.pcscd/drivers/serial
install_pkg/ccid-1.4.28/storage/.kodi/addons/service.pcscd/drivers/serial/libccidtwin.so
install_pkg/ccid-1.4.28/storage/.kodi/addons/service.pcscd/drivers/ifd-ccid.bundle
install_pkg/ccid-1.4.28/storage/.kodi/addons/service.pcscd/drivers/ifd-ccid.bundle/Contents
install_pkg/ccid-1.4.28/storage/.kodi/addons/service.pcscd/drivers/ifd-ccid.bundle/Contents/Info.plist
install_pkg/ccid-1.4.28/storage/.kodi/addons/service.pcscd/drivers/ifd-ccid.bundle/Contents/Linux
install_pkg/ccid-1.4.28/storage/.kodi/addons/service.pcscd/drivers/ifd-ccid.bundle/Contents/Linux/libccid.so
```

MilhouseVH replied:
Yes, it's correct. But it's a bit weird. It's unrelated to this PR, of course, but we shouldn't be hard coding `/storage/.kodi/addons/[service|virtual|whatever].[addonid]/` in multiple locations, which is what we seem to be doing. Perhaps something to clean up in a future PR...

MilhouseVH commented on packages/network/iptables/package.mk:
```diff
@@ -8,13 +8,10 @@ PKG_SHA256="993a3a5490a544c2cbf2ef15cf7e7ed21af1845baf228318d5c36ef8827e157c"
 PKG_SITE="http://www.netfilter.org/"
 PKG_URL="http://www.netfilter.org/projects/iptables/files/$PKG_NAME-$PKG_VERSION.tar.bz2"
-PKG_DEPENDS_TARGET="toolchain linux libmnl libnftnl"
+PKG_DEPENDS_TARGET="toolchain linux:host libmnl libnftnl"
```
Nice - this will hopefully allow some packages (connman, avahi, samba etc.) to build a little earlier.

MilhouseVH commented:
Is it worth sending the `PKG_CONFIG_SYSROOT_BASE` patch upstream?

MilhouseVH commented on packages/tools/grub/package.mk:
```diff
@@ -9,6 +9,7 @@ PKG_LICENSE="GPLv3"
 PKG_SITE="https://www.gnu.org/software/grub/index.html"
 PKG_URL="http://git.savannah.gnu.org/cgit/grub.git/snapshot/$PKG_NAME-$PKG_VERSION.tar.gz"
 PKG_DEPENDS_TARGET="toolchain flex freetype:host gnulib:host gettext:host"
```
Happy to say a `grub` rebuild works fine - even with `gnulib:host` dropped from `PKG_DEPENDS_TARGET`. `gnulib` is unpacked correctly by the recursive `PKG_DEPENDS_UNPACK` logic in `scripts/unpack`, so that's all good.

So, I think you just need to drop `gnulib:host`.

dhewg replied:
Ok, will drop gnulib:host

MilhouseVH commented on packages/addons/service/tvheadend42/package.mk:
```diff
@@ -12,6 +12,7 @@ PKG_SITE="http://www.tvheadend.org"
 PKG_URL="https://github.com/tvheadend/tvheadend/archive/$PKG_VERSION.tar.gz"
 PKG_DEPENDS_TARGET="toolchain avahi comskip curl dvb-apps ffmpegx libdvbcsa libhdhomerun \
                     libiconv openssl pngquant:host Python3:host tvh-dtv-scan-tables"
```
Should be OK, you just want to drop the dependency and rely on `PKG_DEPENDS_UNPACK` to do the right thing.

dhewg replied:
Will do, sound right to me

dhewg commented on packages/addons/addon-depends/multimedia-tools-depends/mediainfo/package.mk:
```diff
@@ -8,21 +8,24 @@ PKG_LICENSE="GPL"
 PKG_SITE="http://mediaarea.net/en/MediaInfo/Download/Source"
 PKG_URL="http://mediaarea.net/download/source/mediainfo/${PKG_VERSION}/mediainfo_${PKG_VERSION}.tar.xz"
 PKG_DEPENDS_TARGET="toolchain libmediainfo"
+PKG_DEPENDS_CONFIG="libzen libmediainfo"
```
libmediainfo depends on libzen, so that's covered, but we need both in `PKG_CONFIG_PATH` for the build to succeed

MilhouseVH replied:
OK

dhewg commented on packages/addons/addon-depends/snapcast-depends/asio/package.mk:
```diff
@@ -10,4 +10,11 @@ PKG_URL="https://github.com/chriskohlhoff/asio/archive/asio-$PKG_VERSION.zip"
+
+post_unpack() {
+  mv $PKG_BUILD/asio/* $PKG_BUILD
+}
```
There already is `PKG_CONFIGURE_SCRIPT`, but that didn't work for some reason. It's been ~6 weeks and I don't recall why though :P

MilhouseVH replied:
Yes, in theory that should work but I don't think scripts/autoreconf will work when configure.ac is in a subdirectory - it assumes it will be in the root. So there'd need to be a bit of tweaking to make it work in a sub-directory, but it's probably not worthwhile right now.

dhewg commented on scripts/build:
```diff
@@ -231,6 +231,20 @@ if [ -f "${PKG_CONFIGURE_SCRIPT}" -o -f "${PKG_CMAKE_SCRIPT}" -o -f "${PKG_MESON
 fi
 
 # configure
+if [ -n "${PKG_DEPENDS_CONFIG}" ]; then
```
I'll add `-a -n "${PKG_INSTALL}"`, since the loop just uses `get_install_dir` and that works exactly for those two targets

dhewg commented:
Squashed in the mentioned changes

> Is it worth sending the `PKG_CONFIG_SYSROOT_BASE` patch upstream?

Due to the hardcoded list of dirs to detect a sysroot I don't think it's clean enough for that...

jernejsk commented on projects/Allwinner/devices/A20/bootloader/release:
```diff
@@ -1,9 +1,11 @@
   fi
 
-  cp -a $(get_build_dir linux)/arch/$TARGET_KERNEL_ARCH/boot/dts/sun7i-a20-*.dtb $RELEASE_DIR/3rdparty/bootloader
+  cp -a $SRCDIR/sun7i-a20-*.dtb $RELEASE_DIR/3rdparty/bootloader
```
While you are at it, can you remove these device specific files and make one common? Only additional change is to make last line "cp -a $SRCDIR/*.dtb $RELEASE_DIR/3rdparty/bootloader
". There will be a bit more files, but it doesn't really matter much.

dhewg replied:
Done, with the addition of copying only device specific dtb files (`*-${lowercase device name}-*.dtb`)

MilhouseVH commented on packages/addons/service/locale/package.mk:
```diff
@@ -16,13 +16,13 @@ PKG_ADDON_TYPE="xbmc.service"
+  cp -PR "$(get_install_dir glibc)/usr/share/i18n" \
+         "$(get_install_dir glibc)/.noinstall/charmaps" \
+         "$(get_install_dir glibc)/.noinstall/locales" \
+         "$ADDON_BUILD/$PKG_ADDON_ID"
```
Wow... the way `glibc` deals with `charmaps` and `locales` is a bit of a mess. As such I'm still not entirely sure this is correct!

Previously we were creating the `i18n` directory in the root of the addon, and into the `i18n` directory we copied the `charmaps` and `locales` as sub-directories.

However the code above will copy the `i18n` directory (which will include a `locales` sub-dir, the content of which will depend on `$GLIBC_LOCALES` - it will either be the GLIBC or POSIX locales, and a `UTF-8` only `charmaps` sub-dir for `Generic` only - won't exist for non-Generic) into the root. And then it will also copy the full GLIBC `charmaps` and `locales` into the root of the addon. That's all 3 directories in the root of the addon (with up to 2 sub-dirs in `i18n`), which is a different structure to previously.

I think we only want to create an empty i18n folder, and then into `i18n` copy only the `charmaps` and `locales` directories from `.noinstall` - which will be the full GLIBC `charmaps` and `locales`. We do not want to copy `/usr/share/i18n` as that potentially brings with it the `utf-8` only `charmaps` and/or POSIX-only `locales`.

It's not entirely clear if we should be using the POSIX-only locales in the add-on when `GLIBC_LOCALES` is `no` - my guess is that we probably should, in which case the content of `.noinstall/locales` would need to reflect that.

dhewg replied:
The directory mess wasn't intended, fixed, thanks

MilhouseVH replied:
Sorry I didn't mean that to sound like I was blaming you - I think this is just a historical accumulation of cruft and this PR tipped it over the edge! :)

The latest `locale` and `glibc` changes look good to me, thanks!

dhewg replied:
Don't worry about it, no blaming was involved! If anything I'm annoyed at myself for getting sloppy there ;)

MilhouseVH commented on projects/Allwinner/bootloader/release:
```diff
@@ -0,0 +1,11 @@
+# Copyright (C) 2018-present Team LibreELEC (https://libreelec.tv)
+
+SRCDIR="$BUILD/image/system/usr/share/bootloader"
+
```
Add ` DSTDIR="$RELEASE_DIR/3rdparty/bootloader"` and you can lose the 3x repeated destination paths. :smile:

dhewg replied:
Done, for all bootloader release scripts

dhewg commented:
Added two commits:
* unpack: don't wipe build stamps
* linux: allow module builds after autoremove & unpack

To allows sane image+autoremove builds, followed by builds of kernel packages (which unpack linux again)

dhewg commented:
These addons access `$PKG_BUILD` in `addon()` and require changes to make use of `install_pkg`:
```
packages/addons/driver/sapphire/package.mk
packages/addons/driver/sundtek-mediatv/package.mk
packages/addons/repository/repository.kodinerds/package.mk
packages/addons/repository/repository.linuxserver.docker/package.mk
packages/addons/script/script.config.vdr/package.mk
packages/addons/service/boblightd/package.mk
packages/addons/service/dispmanx_vnc/package.mk
packages/addons/service/docker/package.mk
packages/addons/service/emby/package.mk
packages/addons/service/emby4/package.mk
packages/addons/service/inadyn/package.mk
packages/addons/service/lcdd/package.mk
packages/addons/service/librespot/package.mk
packages/addons/service/minisatip/package.mk
packages/addons/service/oscam/package.mk
packages/addons/service/syncthing/package.mk
packages/addons/service/tigervnc/package.mk
packages/addons/service/tvheadend42/package.mk
packages/addons/service/usbmuxd/package.mk
packages/addons/tools/btrfs-progs/package.mk
packages/addons/tools/dotnet-runtime/package.mk
packages/mediacenter/kodi-binary-addons/audiodecoder.timidity/package.mk
packages/mediacenter/kodi-binary-addons/inputstream.adaptive/package.mk
```

dhewg commented:
* dropped the commit which merged chrome/jre X packages as that doesn't work as intended
* added `LD_LIBRARY_PATH` to the chrome commit as gtk3 failed to build on my updated host toolchain (`libffi.so.7` on the build host and `libffi.so.6` in LE's `$TOOLCHAIN`)
* new patch: document requirements for autoremove: https://github.com/dhewg/LibreELEC.tv/tree/pull/autoremove/packages#directory-structure
